### PR TITLE
Fixes #18234 - display correct icon when no notifications

### PR DESCRIPTION
**Pull request by lizagilman**

### CODE REVIEW

**Comment by tbrisker** [comment]
ok to test

**Comment by ohadlevy** [comment]
@gailsteiger can you please review? - thanks.

File: webpack/assets/javascripts/react_app/components/notifications/NotificationDrawerToggle.test.js [hunk 1]
```diff
@@ -3,34 +3,56 @@ jest.unmock('../../stores/NotificationsStore');
...
 import React from 'react';
-import { shallow } from 'enzyme';
+import { mount, shallow } from 'enzyme';
 import NotificationDrawerToggle from './NotificationDrawerToggle';
 import NotificationsStore from '../../stores/NotificationsStore';
 import NotificationActions from '../../actions/NotificationActions';
 
 function setup() {
-  return shallow(<NotificationDrawerToggle />);
+    return shallow(<NotificationDrawerToggle />);
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/notifications/NotificationDrawerToggle.test.js]: nitpick: please keep the 2 spaces indentation.

File: webpack/assets/javascripts/react_app/stores/NotificationsStore.js [hunk 2]
```diff
@@ -15,6 +15,10 @@ class NotificationsEventEmitter extends AppEventEmitter {
     return (_notifications.data || []);
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/stores/NotificationsStore.js]: nitpick: should we return empty hash vs empty array on line 15?

File: webpack/assets/javascripts/react_app/stores/NotificationsStore.test.js [hunk 3]
```diff
@@ -25,4 +25,10 @@ describe('NotificationsStore', () => {
     expect(result.info[0].id).toBe(18);
     expect(result.info[1].id).toBe(9);
   });
+
+  it('', () => {
+    const result = NotificationsStore.prepareNotifications([]);
+
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/stores/NotificationsStore.test.js]: what does this test do?

**Comment by gailsteiger** [comment]
This is lovely work and thanks for the tests.

However, the [Notification Drawer design](http://www.patternfly.org/pattern-library/communication/notification-drawer/#/_design) calls for the empty icon to be shown when there are no unread notifications, not when there are no notifications at all. (See Note 7)

**Comment by ohadlevy** [comment]
@gailsteiger good catch!

**Comment by tbrisker** [comment]
@gailsteiger I would argue that when there are no notifications at all there are also no unread notifications. I don't think it makes sense to have a full icon when there are no notifications.

**Comment by ohadlevy** [comment]
@tbrisker i read it as we need to check that there are no unread notifications, having no notifications also qualify of course.

**Comment by gailsteiger** [comment]
@tbrisker @ohadlevy There is no contradiction. Empty icon should appear when there are no unread notifications, including when there are no notifications but also when all notifications are read

That was the original intention of my comment.

**Comment by tbrisker** [comment]
@gailsteiger sorry, I misunderstood your previous comment.
@lizagilman - could you please update this to show the correct icon when there are read notifications as well?

File: webpack/assets/javascripts/react_app/stores/NotificationsStore.js [hunk 4]
```diff
@@ -12,7 +12,11 @@ class NotificationsEventEmitter extends AppEventEmitter {
   }
 
   getNotifications() {
-    return (_notifications.data || []);
+    return (_notifications.data || {});
+  }
+
+  getNotificationsCount() {
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/stores/NotificationsStore.js]: probably not needed?

File: webpack/assets/javascripts/react_app/stores/ToastNotificationsStore.js [hunk 5]
```diff
@@ -0,0 +1,107 @@
+import AppDispatcher from '../dispatcher';
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/stores/ToastNotificationsStore.js]: I guess this file is not part of this PR?

File: webpack/assets/javascripts/react_app/components/notifications/NotificationDrawerToggle.js [hunk 6]
```diff
@@ -21,7 +21,7 @@ class NotificationDrawerToggle extends Component {
     switch (actionType) {
       case ACTIONS.RECEIVED_NOTIFICATIONS: {
         this.setState({
-          count: NotificationsStore.getNotifications().length,
+          unreadNotificationsCount: NotificationsStore.hasUnreadNotifications(),
```
Comment (gailsteiger) [inline webpack/assets/javascripts/react_app/components/notifications/NotificationDrawerToggle.js]: calling NotificationsStore.hasUnreadNotifications without arguments
method signature is `hasUnreadNotifications(notifications)`
note also that this only checks unread notifications when notifications are received. Consider the case when there is one unread notification and it is manually marked as read. Then there are no unread notifications but state won't reflect that until next polling request

File: webpack/assets/javascripts/react_app/stores/NotificationsStore.test.js [hunk 7]
```diff
@@ -18,11 +18,29 @@ describe('NotificationsStore', () => {
...
+    expect(result).toBeFalsy();
+  });
+
+  it('returns true when there are unread notifications', () => {
+    const result = NotificationsStore.hasUnreadNotifications(NotificationsDataSeen);
+
+    expect(result).toBeTruthy();
+  });
 });
```
Comment (gailsteiger) [inline webpack/assets/javascripts/react_app/stores/NotificationsStore.test.js]: Both test data sets contain unread notifications, notification where property seen is false
method is not returning correct result

File: webpack/assets/javascripts/react_app/stores/NotificationsStore.js [hunk 8]
```diff
@@ -12,9 +12,13 @@ class NotificationsEventEmitter extends AppEventEmitter {
   }
 
   getNotifications() {
-    return (_notifications.data || []);
+    return (_notifications.data || {});
   }
 
+  hasUnreadNotifications(notifications) {
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/stores/NotificationsStore.js]: shouldnt the function be:
```js
hasUnreadNotifications() {
  return !!(_notifications.find(notification => notification.seen));
}
```

File: webpack/assets/javascripts/react_app/components/notifications/NotificationDrawerToggle.js (hunk 6 above)
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/notifications/NotificationDrawerToggle.js]: I'm a bit puzzled how the tests results did not reflect the missing argument on this function?

Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/components/notifications/NotificationDrawerToggle.js]: @gailsteiger any suggestions/examples on how its best to test updates to component based on store events?

File: webpack/assets/javascripts/react_app/stores/NotificationsStore.js [hunk 9]
```diff
@@ -12,9 +12,13 @@ class NotificationsEventEmitter extends AppEventEmitter {
...
+    return !!(notifications.find(notification => notification.seen));
+
+  }
```
Comment (gailsteiger) [inline webpack/assets/javascripts/react_app/stores/NotificationsStore.js]: function does not return correct result
(a notification is unread if it's seen property is false)

I suggest checking `_notifications` (which is an object and does not support `find`) instead of calling the method with an argument

File: webpack/assets/javascripts/react_app/components/notifications/NotificationDrawerToggle.js (hunk 6 above)
Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/components/notifications/NotificationDrawerToggle.js]: @gailsteiger wont mark as read trigger an update event which will cause the unread counter to be updated anyway?

Comment (gailsteiger) [reply webpack/assets/javascripts/react_app/components/notifications/NotificationDrawerToggle.js]: for the polling use case perhaps it would be best to keep a running count of unread notifications, reset when notifications are received and adjusted when notification is marked as read. The store data for the read notification should then also be updated (my bad)

File: webpack/assets/javascripts/react_app/stores/NotificationsStore.js (hunk 8 above)
Comment (gailsteiger) [reply webpack/assets/javascripts/react_app/stores/NotificationsStore.js]: that function I think returns true if any of the notifications has its seen property set to true

File: webpack/assets/javascripts/react_app/components/notifications/NotificationDrawerToggle.js (hunk 6 above)
Comment (lizagilman) [reply webpack/assets/javascripts/react_app/components/notifications/NotificationDrawerToggle.js]: @ohadlevy @gailsteiger 
will go over everything in this PR and fix according to your comments!

Comment (gailsteiger) [reply webpack/assets/javascripts/react_app/components/notifications/NotificationDrawerToggle.js]: @lizagilman great. Have a look at NotificationPanel.js Consider extracting the calculation of the unread notifications to a helper or services file and reuse for your purposes.

```js
    let unread;

    if (notifications) {
      unread = notifications.reduce((total, curr) => {
        if (!curr.seen) {
          total = total + 1;
        }
        return total;
      }, 0);
    } else {
      unread = 0;
    }
```

File: webpack/assets/javascripts/react_app/components/notifications/NotificationDrawerToggle.test.js [hunk 10]
```diff
@@ -18,6 +18,14 @@ describe('NotificationDrawerToggle', () => {
...
   });
 
+  it('should show bell icon', () => {
+    const wrapper = setup();
+    const emptyBellIcon = wrapper.find('.fa-bell-o');
+    const fullBellIcon = wrapper.find('.fa-bell');
+
+    expect(emptyBellIcon.length || fullBellIcon.length).toBe(1);
+  });
+
```
Comment (lizagilman) [inline webpack/assets/javascripts/react_app/components/notifications/NotificationDrawerToggle.test.js:28]: overall tests will be followed when finished moving from flux to redux

**Comment by ohadlevy** [comment]
@matanwerbner @gailsteiger can you please review? thanks

**Comment by gailsteiger** [comment]
@lizagilman I added some unread messages to the database and css class is still `fa-bell-o`

File: webpack/assets/javascripts/react_app/components/notifications/NotificationDrawerToggle.js [hunk 11]
```diff
@@ -25,6 +26,8 @@ class NotificationDrawerToggle extends Component {
       case ACTIONS.RECEIVED_NOTIFICATIONS: {
         this.setState({
           count: NotificationsStore.getNotifications().length,
+          unreadCount: _.filter(NotificationsStore.getNotifications(),
+            notification => notification.seen).length,
           isLoaded: true
         });
         break;
```
Comment (gailsteiger) [inline webpack/assets/javascripts/react_app/components/notifications/NotificationDrawerToggle.js]: I've found that it is best to make the `setState` call as simple as possible. I would remove the function calls from the `setState` call, do all the calculations before and pass a simple object literal.

Also, don't call `NotificationsStore.getNotifications()` twice. Cache the result of the first call.

`NotificationsStore.getNotifications()` returns an object of arrays of notifications. You have to do your calculations on the underlying arrays.

Comment (matanwerbner) [reply webpack/assets/javascripts/react_app/components/notifications/NotificationDrawerToggle.js]: I think this should be !seen...we both missed it lol

Comment (gailsteiger) [reply webpack/assets/javascripts/react_app/components/notifications/NotificationDrawerToggle.js]: All the more reason not to do the calculations in `setState`. Kinda hard to set a break point.

Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/components/notifications/NotificationDrawerToggle.js]: anyway we can add a test for this ? :)

Comment (matanwerbner) [reply webpack/assets/javascripts/react_app/components/notifications/NotificationDrawerToggle.js]: i dont think there is a way to test flux connected components.. @gailsteiger  do you know a way to mock the store?

**Comment by lizagilman** [comment]
@gailsteiger can you please review again? thanks.

File: webpack/assets/javascripts/react_app/components/notifications/NotificationDrawerToggle.js [hunk 12]
```diff
@@ -23,10 +24,18 @@ class NotificationDrawerToggle extends Component {
   onChange(actionType) {
     switch (actionType) {
       case ACTIONS.RECEIVED_NOTIFICATIONS: {
-        this.setState({
-          count: NotificationsStore.getNotifications().length,
-          isLoaded: true
-        });
+        const notifications = NotificationsStore.getNotifications();
+
+        const newState = _.reduce(notifications, (curr, next) => {
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/notifications/NotificationDrawerToggle.js]: maybe i'm missing something, but why do we care about the current count ? just set the new state regardless ?

Comment (lizagilman) [reply webpack/assets/javascripts/react_app/components/notifications/NotificationDrawerToggle.js]: we might need it for later use @matanwerbner ?

Comment (matanwerbner) [reply webpack/assets/javascripts/react_app/components/notifications/NotificationDrawerToggle.js]: I think that If we should need it in the future, we will implement. Currently, if its not needed, it should be removed

Comment (matanwerbner) [reply webpack/assets/javascripts/react_app/components/notifications/NotificationDrawerToggle.js]: All this component needs is a bool called "hasUnreadMessages" imho 😃

Comment (gailsteiger) [reply webpack/assets/javascripts/react_app/components/notifications/NotificationDrawerToggle.js]: I think it is important to write some tests that will check for various scenarios, including a scenario where there is more than 1 array of notifications.

Comment (matanwerbner) [reply webpack/assets/javascripts/react_app/components/notifications/NotificationDrawerToggle.js]: @gailsteiger i agree, we're a bit lost on testing UI components with flux, do you have any idea how to do this?
in redux it would much simpler as the store is injected into the component, so different data can be mocked in the store.
in flux currently nothing is injected so there is much less control.

**Comment by lizagilman** [comment]
@ohadlevy @gailsteiger can you please review again?

**Comment by ohadlevy** [comment]
thanks @lizagilman
